### ISSUE TITLE
Improve UI/UX of bookmark popup

### ISSUE THREAD

**Issue by bradleyrichter** [brave/browser-laptop#2894]
The current bookmark popup appears in the center of the BM bar and has that "not finished" look.

![image](https://cloud.githubusercontent.com/assets/13509546/17345985/8c1970e6-58bd-11e6-861c-40f5b6cede2f.png)

Let's improve the styling and UX with this design:

![image](https://cloud.githubusercontent.com/assets/13509546/17354901/781d60cc-5901-11e6-99e1-d3ee201f89fd.png)

- The editable field for the path should not be necessary as the user can either edit it before pressing the BM button, or edit it later. (we can gauge feedback or add a method later)

- the name of the BM should be taken from the page title (if existing) and open as selected so typing can begin

- the bookmark is created when the Add button is pressed, but this dialog allows the user to append changes in the form of the text or location
- if the location is changed, the BM must actually be moved
- remove will remove the BM, uncheck the BM icon and close the dialog
- Done accepts and closed the dialog with info as shown
- close-button = done
- the default folder is the Bookmarks toolbar
- a "View all bookmarks" link/button on the bottom provides a shortcut to the bookmarks manager. (about:bookmarks)
- the popup should open at the location of the bookmark button

- address keyboard navigation flow as described below by @neeklamy 
- recall last use folder after initial default to Bookmarks Bar

(note: the UI behind the popup in this image will be specified in a separate issue)

Part 2 of this UI task is to extend the component for use as the bookmark edit dialog. It should appear at the actual bookmark which is being edited on the BM bar.

![image](https://cloud.githubusercontent.com/assets/13509546/17796934/5a5c4128-6579-11e6-8441-4f7947f897bb.png)

- size extends to make room for the URL field
- title changes to "Edit Bookmark"
- button text changes to "Cancel/Save"

cc @bsclifton

**Comment by gulkily** [comment]
I disagree that the editable field for the path is not necessary. I clean up the URL I'm adding quite often, and get quite annoyed at browsers that don't allow this in the add-bookmark dialog.

If the field is not there, then I have to break my flow to find the bookmark, wherever it may be, open the edit dialog, and only then edit and save it.

**Comment by bradleyrichter** [comment]
@gulkily Thanks for the honest feedback. It does seem that browsers vary in this regard.

@bbondy Is there any way to allow the URL field to stay editable while the bookmark hanger is open? I am trying to avoid the complete duplication of the URL string, and keep the popup minimal and easy.

There may be a side effect that would need quick testing.

**Comment by gulkily** [comment]
One more thought:

>no bookmark is actually made until the Add button is pressed

This goes against the traditional add-bookmark action that is triggered by the "Bookmark Page" or "Add Bookmark" menu item usually bound to Ctrl+D/Cmd+D.

So changing this behavior will either require a different behavior between "click the star button" and "use the menu item/keyboard shortcut" or break a standard workflow that goes back to the 90s.

**Comment by bradleyrichter** [comment]
Yep. Let's stay with history, and iterate elsewhere to improve bookmarking in general.

**Comment by bradleyrichter** [comment]
If anyone is curious, here is how the other browsers do it:

![image](https://cloud.githubusercontent.com/assets/13509546/17354832/d1482728-5900-11e6-973a-a39c9710d6db.png)

Chrome, FIrefox and Opera set the bookmark and let you undo it.

Safari and VIvaldi seem to treat it like a dialog that can be canceled.

Make sense to me to follow the leaders on this one.


Also, Chrome allows editing the URL in a second, less organized dialog.

Vivaldi might allow it inline. The rest don't seem to allow it.

**Comment by neeklamy** [comment]
Some thoughts.

- On the default folder, I *really like* how Safari and Google Chrome both remember the last used folder.
- Keyboard navigation would be nice too, so a workflow like this:  
  1. Command-D,
  2. Edit the bookmark name,
  3. <kbd>Tab</kbd>
  4. Type first few letters of a different folder, folder is selected,
  5. <kbd>Enter</kbd>
  6. Popup closes. Hitting <kbd>Enter</kbd> at any point, whichever input is key, should finalise the bookmark and close the popup.

**Comment by bradleyrichter** [comment]
@neeklamy Great addition. Thanks for the descriptive steps. I'll note them above.

**Comment by rohanthacker** [comment]
This UI looks great!

**Comment by bradleyrichter** [comment]
@jkup Jon - it would be great if you can look at @rohanthacker 's fork and refine/merge as needed.

cc @bbondy

**Comment by jkup** [comment]
Happy to!

**Comment by bradleyrichter** [comment]
I added a panel variation to be used for editing boookmarks from the BM bar. ^^

**Comment by bbondy** [comment]
The edit bookmark variation should also be used when adding a bookmark not for a specific page, for example right click the bookmarks toolbar and add bookmark. I think in that case it would be centered though.

**Comment by rohanthacker** [comment]
@jkup Hey Jon - This is the link to the [changes](https://github.com/brave/browser-laptop/compare/master...rohanthacker:improved-bookmarks-ui) 

As of now I've not been able to figure out how icons are used, and how to updated the Strings that are used in the l10n prop.

**Comment by lucidNTR** [comment]
i hope i did not miss this but i would really like to only see the enabled delete button when the page is actually a bookmark. this confuses me in the current master behavior.

**Comment by bradleyrichter** [comment]
@lucidNTR That will happen with the new design. Thanks for the note.

**Comment by jkup** [comment]
Hey @rohanthacker this looks great! Any chance you can rebase your commit and submit a PR? I'll happy test it out and merge it in!

**Comment by rohanthacker** [comment]
@jkup I'd be happy to however, I hit a small roadblock with the drop-down menu. Please do see #3282

**Comment by bbondy** [comment]
@jkup I'm not sure if you're finishing off some ledger stuff, but otherwise pls work on this in the 0.12.6 timeframe. Thanks!

**Comment by srirambv** [comment]
@bradleyrichter  part 2 of what you have mentioned is not how the current behavior is 

![image](https://cloud.githubusercontent.com/assets/17010094/19597060/e1410634-97af-11e6-8274-4666b14a1dda.png)

This is the current behavior
![bookmarkhanger](https://cloud.githubusercontent.com/assets/17010094/19597086/0fdae05a-97b0-11e6-803d-7b55ef6c6bc5.gif)

**Comment by bradleyrichter** [comment]
since the code was using the same dialog for all cases, we now need to make a new centered modal to accommodate the other uses.

when adding a folder:
![image](https://cloud.githubusercontent.com/assets/13509546/19671255/78347a7a-9a3b-11e6-8f0f-028209bba663.png)



when adding a new BM when no URL is present to start with:
![image](https://cloud.githubusercontent.com/assets/13509546/19671279/a445bdf4-9a3b-11e6-915d-ca844a94507d.png)


So to summarize, the door hanger should only open at the BM star button or at the BM toolbar when it is relevant. Otherwise, it should open as a modal with correct titles based on activity.

**Comment by bsclifton** [comment]
Solution provided in https://github.com/brave/browser-laptop/pull/5118 doesn't quite look this good, but it's close :smile: (screenshots posted in the PR)

**Comment by bradleyrichter** [comment]
@jkup you mentioned this needs more time. Did you have an estimate? LMK if we should track it separately in asana.

**Comment by bsclifton** [comment]
@bradleyrichter what's left? do you mean the additional styles you show in the screen caps above? (blue around text box, down arrow under the name box, etc)
